### PR TITLE
WT-10007 Create function to remove local file with fh_lock.

### PR DESCRIPTION
**Pull request by sueloverso**
I moved code into os_fhandle.c because functions that take the conn->fh_lock seem to all live in there. I also considered, instead, adding something like WT_WITH_FHLOCK() to wrap a function. That would have been a more complex change and this area the only user of it.

### CODE REVIEW

File: src/os_common/os_fhandle.c [hunk 1]
```diff
@@ -72,6 +72,33 @@ __wt_handle_is_open(WT_SESSION_IMPL *session, const char *name)
     return (found);
 }
 
+/*
+ * __wt_remove_tiered_local --
+ *     While locked, if the handle is not open, remove the local object.
+ */
+int
+__wt_remove_tiered_local(WT_SESSION_IMPL *session, const char *object, bool *removed)
```
Comment (keitharnoldsmith) [inline src/os_common/os_fhandle.c:80]: I don't think the name of this function should embed the reason we (currently) use it. It's just a function that safely removes a file in a way that avoids races with another thread trying to open the file. We use it for removing local tiered objects today, but we might find other uses for it in the future. 

I would change the name of this function to something more generic. Maybe `__wt_lock_and_remove()`? or `__wt_atomic_remove()`?

**Review by keitharnoldsmith — COMMENTED** [review]
It looks like some of the unit tests timed out. I'm guessing that's because of the deadlock mentioned in one of my comments.

Thinking about this problem a small bit, I wonder if we should be addressing this at a higher level. We now have a separate WT_BLOCK for each file/object. So maybe the we should remove the local file by setting a "remove-on-close" flag on it's WT_BLOCK and using that to delete the file when the WT_BLOCK is closed.

Happy to brainstorm about this either now or at the Tora meeting.

File: src/os_common/os_fhandle.c (hunk 1 above)
Comment (sueloverso) [reply src/os_common/os_fhandle.c:80]: I like the generic naming suggestions.

File: src/os_common/os_fhandle.c [hunk 2]
```diff
@@ -72,6 +72,33 @@ __wt_handle_is_open(WT_SESSION_IMPL *session, const char *name)
...
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+
+    conn = S2C(session);
+    *removed = false;
+    __wt_spin_lock(session, &conn->fh_lock);
```
[the remark this answers is not in the document]
Comment (sueloverso) [reply src/os_common/os_fhandle.c:87]: Somehow my testing didn't hit this. Will fix.

**Review by ddanderson — APPROVED** [review]
LGTM

**Review by keitharnoldsmith — APPROVED** [review]
LGTM. 

While it seems to me that the recent changes should address the deadlock issue, there is an Evergreen failure that looks like a timeout in the Python tests. This message looks suspicious:
```
[2023/07/19 23:10:04.701] [1689804246:927732][71962:0x7f106e758700], test_tiered04.test_tiered04.test_tiered(dir_store), sweep-server: [WT_VERB_DEFAULT][WARNING]: Session 14 (@: 0x0x3a92280 name: WT_SESSION.checkpoint) did not run a sweep for 60 minutes.
```
I assume `test_tiered04` shouldn't run for over 60 minutes.

**Comment by sueloverso** [comment]
Yes, I just got back to this PR and noticed the timeouts. I downloaded the artifacts but haven't looked at it yet. It does seem like there is still some kind of issue.

**Comment by sueloverso** [comment]
@keitharnoldsmith I fixed the deadlock but it touched a few files due to adding a function arg. I've been considering the block flag suggestion you made. We can chat about it if you want at the next Tora meeting.

**Review by keitharnoldsmith — APPROVED** [review]
LGTM

**Pull request merged**